cookiejar.go

```diff
@@ -0,0 +1,205 @@
+	cj.m.Lock()
+	{
+		cookies = cj.hostCookies[hostStr]
+		var (
+			t = time.Now()
+			n = len(cookies)
+		)
+		for i := 0; i < len(cookies); i++ {
+			c := cookies[i]
+			if !c.Expire().Equal(CookieExpireUnlimited) && c.Expire().Before(t) { // cookie expired
+				cookies = append(cookies[:i], cookies[i+1:]...)
+				ReleaseCookie(c)
+				i--
+			}
+		}
+		// has any cookie been deleted?
+		if n > len(cookies) {
+			cj.hostCookies[hostStr] = cookies
+		}
+	}
+	cj.m.Unlock()
+	return
+}
+
+// Set sets cookies for a specific host.
+//
+// The host is get from uri.Host().
+//
+// If the cookie key already exists it will be replaced by the new cookie value.
+//
+// CookieJar keeps a copy of the cookies, so the parsed cookies can be released safely.
+func (cj *CookieJar) Set(uri *URI, cookies ...*Cookie) {
+	if uri != nil {
+		cj.set(uri.Host(), cookies...)
+	}
+}
+
+// SetByHost sets cookies for a specific host.
+//
+// If the cookie key already exists it will be replaced by the new cookie value.
+//
+// CookieJar keeps a copy of the cookies, so the parsed cookies can be released safely.
+func (cj *CookieJar) SetByHost(host []byte, cookies ...*Cookie) {
+	cj.set(host, cookies...)
+}
+
+func (cj *CookieJar) set(host []byte, cookies ...*Cookie) {
+	cj.m.Lock()
+	{
+		if cj.hostCookies == nil {
+			cj.hostCookies = make(map[string][]*Cookie)
+		}
+		hostStr := b2s(host)
+		hcs, ok := cj.hostCookies[hostStr]
+		if !ok {
+			// If the key does not exists in the map then
+			// we must make a copy for the key.
+			hostStr = string(host)
+		}
+		for _, cookie := range cookies {
+			c := searchCookieByKeyAndPath(cookie.Key(), cookie.Path(), hcs)
+			if c == nil {
+				c = AcquireCookie()
+				hcs = append(hcs, c)
+			}
+			c.CopyTo(cookie)
+		}
+		cj.hostCookies[hostStr] = hcs
+	}
+	cj.m.Unlock()
+}
+
+// SetKeyValue sets a cookie by key and value for a specific host.
+//
+// This function prevents extra allocations by making repeated cookies
+// not being duplicated.
+func (cj *CookieJar) SetKeyValue(host, key, value string) {
+	cj.SetKeyValueBytes(host, s2b(key), s2b(value))
+}
+
+// SetKeyValueBytes sets a cookie by key and value for a specific host.
+//
+// This function prevents extra allocations by making repeated cookies
+// not being duplicated.
+func (cj *CookieJar) SetKeyValueBytes(host string, key, value []byte) {
+	cj.setKeyValue(host, key, value)
+}
+
+func (cj *CookieJar) setKeyValue(host string, key, value []byte) {
+	c := AcquireCookie()
+	c.SetKeyBytes(key)
+	c.SetValueBytes(value)
+	cj.set(s2b(host), c)
+}
+
+func (cj *CookieJar) dumpTo(req *Request) {
+	uri := req.URI()
+	cookies := cj.get(uri.Host(), uri.Path())
+	for _, cookie := range cookies {
+		req.Header.SetCookieBytesKV(cookie.Key(), cookie.Value())
+	}
+}
+
+func (cj *CookieJar) getFrom(host, path []byte, res *Response) {
+	hostStr := b2s(host)
+	cj.m.Lock()
+	{
+		if cj.hostCookies == nil {
+			cj.hostCookies = make(map[string][]*Cookie)
```

Comment:
Here you can just do `return`. No need to allocate anything.

Reply:
Just make sure you do `defer cj.m.Unlock()` instead of doing this at the end of the function.

I myself always prefer to `defer` `Unlock`s as its much easier to follow the code. Also `defer` is basically free these days so no worry about performance.